core/src/graph/page.rs

```diff
@@ -0,0 +1,411 @@
+use crate::{
+	dsnp::{api_types::*, dsnp_types::*},
+	util::time::time_in_ksecs,
+};
+use anyhow::{Error, Result};
+
+use crate::dsnp::{
+	compression::{CompressionBehavior, DeflateCompression},
+	encryption::EncryptionBehavior,
+	schema::SchemaHandler,
+};
+
+const APPROX_MAX_CONNECTIONS_PER_PAGE: usize = 10; // todo: determine best size for this
+
+/// Graph page structure
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct GraphPage {
+	/// Page ID
+	page_id: PageId,
+	/// Privacy type of owning graph
+	privacy_type: PrivacyType,
+	/// Current content hash of page as retrieved from chain
+	content_hash: u32,
+	/// List of PRIds
+	prids: Vec<DsnpPrid>,
+	/// List of connections
+	connections: Vec<DsnpGraphEdge>,
+}
+
+/// Conversion for Public Graph
+impl TryFrom<&PageBlob> for GraphPage {
+	type Error = Error;
+	fn try_from(PageBlob { content_hash, payload, page_id, .. }: &PageBlob) -> Result<Self> {
+		let DsnpUserPublicGraphChunk { compressed_public_graph } =
+			DsnpUserPublicGraphChunk::try_from(payload.as_slice())?;
+		let uncompressed_chunk = DeflateCompression::decompress(&compressed_public_graph)?;
+		Ok(Self {
+			page_id: *page_id,
+			privacy_type: PrivacyType::Public,
+			content_hash: *content_hash,
+			prids: Vec::new(),
+			connections: SchemaHandler::read_inner_graph(&uncompressed_chunk)?,
+		})
+	}
+}
+
+/// Conversion for Private Graph
+impl<E: EncryptionBehavior> TryFrom<(&PageBlob, &Vec<KeyPair<E>>)> for GraphPage {
+	type Error = Error;
+	fn try_from(
+		(PageBlob { content_hash, payload, page_id, .. }, keys): (&PageBlob, &Vec<KeyPair<E>>),
+	) -> Result<Self> {
+		let DsnpUserPrivateGraphChunk { key_id, encrypted_compressed_private_graph, prids } =
+			DsnpUserPrivateGraphChunk::try_from(payload.as_slice())?;
+		let mut decrypted_chunk: Option<Vec<u8>> = None;
+
+		// First try the key that was indicated in the page
+		if let Some(indicated_key) = keys.iter().find(|k| k.key_id == key_id) {
+			if let Ok(data) =
+				E::decrypt(&encrypted_compressed_private_graph, &indicated_key.private_key)
+			{
+				decrypted_chunk = Some(data);
+			}
+		}
+
+		// If we couldn't decrypt with (or find) the indicated key, try all keys
+		decrypted_chunk = match decrypted_chunk {
```

Comment:
I'm still skeptical if we should try with other keys since it might be successful even though the data might be garbage. I think it depends on poly1305. Does it always fail if wrong key used to decrypt to wrong data